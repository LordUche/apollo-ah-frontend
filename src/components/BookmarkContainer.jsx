/* eslint-disable react/jsx-indent */
/* eslint-disable react/jsx-closing-tag-location */
/* eslint-disable react/no-array-index-key */
/* eslint-disable react/jsx-wrap-multilines */
import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import BookmarkItem from '../views/BookmarkItem';
import { getBookmarkedArticles } from '../actions/getBookmark';
import * as selectors from '../selectors/getBookmarkSelector';

/**
 * @description - returns all articles bookmarked by a user
 * @return {JSX}
 */
function BookmarkContainer(props) {
  const [viewType, setViewType] = useState('row view-row');

  const { articles } = props;

  const getArticlesHandler = () => {
    props.getBookmarkedArticles();
  };

  const changeViewToHorizontal = () => {
    setViewType('row view-row list-view');
  };

  const changeViewToVertical = () => {
    setViewType('row view-row');
  };

  useEffect(() => (
    getArticlesHandler()
  ), []);

  return (
    <>
      <header className="page-header" data-pg-collapsed>
        <h1>Bookmarks </h1>
        <div className="btn-group btn-group-toggle float-right view-changer" data-toggle="buttons">
          <label className="btn active" onClick={changeViewToVertical}>
            <input type="radio" name="options" autoComplete="off" checked value="" />
            <i className="fas fa-th" />
          </label>
          <label className="btn" onClick={changeViewToHorizontal}>
            <input type="radio" name="options" autoComplete="off" value="list-view" />
            <i className="fas fa-list" />
          </label>
        </div>
      </header>
      {
        articles.articles.length !== 0
          ? (
            <>
              <div className={viewType}>
                {
                  // eslint-disable-next-line arrow-body-style
                  articles.articles.map((article, indx) => {
                    return (
                      <BookmarkItem
                        key={indx}
                        category={article.Article.articleCategory.category}
                        image={article.Article.image}
                        articleTitle={article.Article.title}
                        articleDescription={article.Article.description}
                        authorsName={article.Article.User.Profile.firstname}
                        authorsImage={article.Article.User.Profile.image}
                        readTime={article.Article.readTime}
                      />
                    );
                  })
                }
              </div>
              {/** pagination support to be implemented later */}
              <div>Pagination is here</div>
            </>
          ) : <div>
            <i className="fas fa-book-open" style={{ fontSize: '70px', color: '#66008c' }} />
            <p style={{ paddingTop: '30px' }}>*It looks like you have not bookmarked any article yet</p>
          </div>
      }
    </>
  );
}

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getBookmarkLoading,
    articles: selectors.getBookmarkArticles,
  }
);

export default connect(mapStateToProps, { getBookmarkedArticles })(BookmarkContainer);
