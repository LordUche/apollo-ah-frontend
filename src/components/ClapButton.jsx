/* eslint-disable no-undef */
import React, { Component } from 'react';
import PropType from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { debounce } from 'lodash';
import { withRouter } from 'react-router-dom';
import clapIcon from '../images/clap.svg';
import * as articleSelector from '../selectors/singleArticleSelector';
import { getAuthId, getIsAuthenticated } from '../selectors/authSelector';
import { getToken } from '../selectors/profileSelector';
import { clapArticleRequest, fetchUserClaps } from '../actions/clapsAction';
import { getUserClaps } from '../selectors/clapsSelector';
import { CLAPS_LIMIT, CLAPS_REQUEST_TIMER } from '../utils/contants';

/**
 * A class representing clap button component
 *
 * @class ClapButton
 * @extends {Component}
 */
export class ClapButton extends Component {
  /**
   * Handles clap request (debounced)
   *
   * @returns {void}
   * @memberof ClapButton
   */
  makeClapRequest = debounce(() => {
    const { clapArticle, articleSlug: slug, token } = this.props;
    const { clapsCount: claps } = this.state;
    this.setState({ clapsCount: 0 });
    return clapArticle({ slug, claps, token });
  }, CLAPS_REQUEST_TIMER);

  static propTypes = {
    articleSlug: PropType.string.isRequired,
    articleAuthorId: PropType.number.isRequired,
    articleClaps: PropType.number.isRequired,
    clapArticle: PropType.func.isRequired,
    isLoggedIn: PropType.bool.isRequired,
    userId: PropType.number,
    history: PropType.object.isRequired,
    token: PropType.string.isRequired,
    userClaps: PropType.number.isRequired,
    loadUserClaps: PropType.func.isRequired,
  };

  static defaultProps = {
    userId: null,
  };

  constructor(props) {
    super(props);
    this.state = {
      clapsLeft: CLAPS_LIMIT - props.userClaps,
      articleClaps: props.articleClaps,
      clapsCount: 0, // record clap button clicks
    };
  }

  componentDidMount = async () => {
    const {
      articleSlug: slug,
      isLoggedIn,
      userId,
      loadUserClaps
    } = this.props;
    if (isLoggedIn) {
      loadUserClaps({ slug, userId });
    }
  }

  /**
   * Makes an article clap
   *
   * @returns {void}
   * @memberof ClapButton
   */
  handleClapArticle = () => {
    const {
      isLoggedIn,
      userId,
      articleAuthorId,
      history,
    } = this.props;

    if (!isLoggedIn) { // redirect guest users to login
      history.push('/login');
      return false;
    }
    // Deny article's author
    if (articleAuthorId === userId) return false;
    const { articleClaps, clapsLeft, clapsCount } = this.state;
    if (clapsLeft >= 1) {
      this.setState({
        clapsLeft: clapsLeft - 1,
        articleClaps: articleClaps + 1,
        clapsCount: clapsCount + 1,
      });
      this.makeClapRequest();
    }
  }

  /**
   * React render function
   *
   * @returns {JSX.Element} DOM element
   * @memberof ClapButton
   */
  render() {
    const { articleClaps } = this.state;

    return (
      <span className="clap-icon">
        <button type="button" className="btn" onClick={this.handleClapArticle}>
          <img src={clapIcon} alt="" />
        </button>
        {articleClaps}
        {' '}
        claps
      </span>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  articleSlug: articleSelector.getArticleSlug,
  articleAuthorId: articleSelector.getArticleAuthorId,
  articleClaps: articleSelector.getArticleClaps,
  isLoggedIn: getIsAuthenticated,
  userId: getAuthId,
  token: getToken,
  userClaps: getUserClaps,
});

const mapDispatchToProps = {
  clapArticle: claps => clapArticleRequest(claps),
  loadUserClaps: payload => fetchUserClaps(payload),
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(ClapButton));
