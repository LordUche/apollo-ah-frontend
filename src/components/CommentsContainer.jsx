/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';

import Comments from '../views/Comments';

/**
 * Container component for the Article comments view
 * @export
 * @class CommentsContainer
 * @extends {Component}
 */
export default class CommentsContainer extends Component {
  render() {
    return (
      <Comments />
    );
  }
}
