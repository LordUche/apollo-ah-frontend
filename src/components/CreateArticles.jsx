/* eslint-disable no-unused-vars */
/* eslint-disable arrow-parens */
/* eslint-disable react/jsx-no-bind */
/* eslint-disable jsx-a11y/no-redundant-roles */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable jsx-a11y/label-has-for */
import React, { useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import * as selectors from '../selectors/createArticleSelector';
import createArticleAction from '../actions/createArticles';
import CreateArticle from '../views/CreateArticle';
import CloudinaryWidget from './common/CloudinaryWidget';


/**
 * Form for creating a new article
 * @param {object} props Component props
 * @param {function} props.createArticleAction action creator function
 */
function CreateArticleContainer(props) {
  const [formData, setFormData] = useState({
    title: null,
    description: null,
    body: null,
    categoryId: null,
    tagList: [],
    image: null,
  });

  /** passing global state loading as props into variables */
  let { message, isLoading, token } = props;

  let widget;

  const updateInputHandler = (event) => {
    setFormData({ ...formData, [event.target.name]: event.target.value });
  };

  const updateBodyHandler = (value) => {
    setFormData({ ...formData, body: value });
  };

  const updateTagHandler = (tags) => {
    const listOfTags = [];
    tags.forEach(tag => {
      listOfTags.push(tag.text);
      setFormData({ ...formData, tagList: listOfTags });
    });
  };

  const submitHandler = (event) => {
    event.preventDefault();
     console.log('--->', formData);
     const requestData = {
       formData,
       token,
     };
    //props.createArticleAction(formData);
    props.createArticleAction(requestData);
  };

  const cloudinaryUpdate = (imageUrl) => {
    setFormData({ ...formData, image: imageUrl });
  };

  const imageUploadHandler = () => {
    // console.log('yo mama');
    // const { showLoader } = this.props;
    // showLoader(true);
    if (widget === undefined) {
      widget = new CloudinaryWidget(
        cloudinaryUpdate, (error) => {
        return (error);
      }, false);
    }
    widget.open();
  };

  console.log('--->', isLoading);
  console.log('==->', message);

  return (
    <CreateArticle
      updateInputHandler={updateInputHandler}
      updateBodyHandler={updateBodyHandler}
      updateTagHandler={updateTagHandler}
      imageUploadHandler={imageUploadHandler}
      submitHandler={submitHandler}
      valueText={formData.body}
    />
  );
}

const mapStateToProps = createStructuredSelector(
  {
    isLoading: selectors.getCreateArticleLoadingState,
    message: selectors.getCreateArticleMessageState,
    token: selectors.getUserToken,
  }
);

export default connect(mapStateToProps, { createArticleAction })(CreateArticleContainer);
