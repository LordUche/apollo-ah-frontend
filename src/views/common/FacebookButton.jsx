import React from 'react';
import FacebookLogin from 'react-facebook-login/dist/facebook-login-render-props';
import PropTypes from 'prop-types';
import facebook from '../../images/icons/facebook.svg';
import ImageButton from './ImageButton';

/**
 * A function representing FacebookButton
 *
 * @param {{Function}} handleResponse - A response handler callback
 * @returns {JSX.Element} - DOM element
 */
const FacebookButton = ({ handleResponse }) => (
  <FacebookLogin
    appId={process.env.FACEBOOK_APP_ID}
    autoLoad={false}
    fields="name,email,picture"
    callback={handleResponse}
    render={renderProps => (
      <ImageButton
        imgSrc={facebook}
        onClick={renderProps.onClick}
        altText="login with facebook"
      />
    )}
  />
);

FacebookButton.propTypes = {
  handleResponse: PropTypes.func.isRequired,
};

export default FacebookButton;
