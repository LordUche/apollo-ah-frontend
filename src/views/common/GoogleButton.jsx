import React from 'react';
import PropTypes from 'prop-types';
import GoogleLogin from 'react-google-login';
import google from '../../images/icons/google.svg';
import ImageButton from './ImageButton';

/**
 * A function representing GoogleButton
 *
 * @param {{Function}} handleResponse - A response handler callback
 * @returns {JSX.Element} - DOM element
 */
const GoogleButton = ({ handleResponse }) => {
  return (
    <GoogleLogin
      clientId={process.env.GOOGLE_CLIENT_ID}
      autoLoad={false}
      render={renderProps => (
        <ImageButton
          imgSrc={google}
          onClick={renderProps.onClick}
          altText="login with google"
        />
      )}
      onSuccess={handleResponse}
      onFailure={handleResponse}
    />
  );
};

GoogleButton.propTypes = {
  handleResponse: PropTypes.func.isRequired,
};

export default GoogleButton;
