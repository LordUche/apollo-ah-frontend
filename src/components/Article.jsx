import React from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import ArticleBody from '../views/ArticleBody';
import CommentsContainer from './CommentsContainer';

/**
 *
 *
 * @class Article
 * @extends {React.Component}
 */
class Article extends React.Component {
  /**
   *
   * @static - Validating proptypes
   * @memberof Article
   */
  static propTypes = {
    match: PropTypes.object.isRequired,
  }

  /**
   * Creates an instance of Article.
   * @param {*} props - Props passed down from the BrowserRouter
   * @memberof Article
   */
  constructor(props) {
    super(props);
    this.state = {
      article: {},
      bookmarked: false,
    };
  }

  /**
   *
   * @todo - Move the fetch article functionality to redux store
   * @memberof Article
   */
  componentDidMount = async () => {
    const { match } = this.props;
    const { slug } = match.params;
    const result = await axios.get(`${process.env.API_BASE_URL}/articles/${slug}`);
    this.setState({ article: result.data.data });
  }

  /**
   * @todo - employ javascript debounce to avoid calling API on repeated clicks
   * @method bookmarkArticle - The bookmark article functionality
   * @memberof Article
   */
  bookmarkArticle = () => {
    const { bookmarked } = this.state;
    this.setState({ bookmarked: !bookmarked });
  };

  /**
   *
   *
   * @returns {JSX} The JSX representation of the Article component
   * @memberof Article
   */
  render() {
    const { article, bookmarked } = this.state;
    return (
      <div>
        <main className="main-body">
          <section className="min-vh-100">
            {article.image && (
              <div
                className="pg-empty-placeholder single-hero"
                style={{ backgroundImage: `url(${article.image})` }}
              />
            )}
            <div className="container">
              <div className="single-container">
                <ArticleBody
                  article={article}
                  bookmarkArticle={this.bookmarkArticle}
                  bookmarked={bookmarked}
                />
                <CommentsContainer />
              </div>
            </div>
            <div className="single-suggested-grp">
              <div className="container">
                <h4>Recommended for you</h4>
                <div className="row pg-empty-placeholder" />
              </div>
            </div>
          </section>
        </main>
      </div>
    );
  }
}

export default Article;
