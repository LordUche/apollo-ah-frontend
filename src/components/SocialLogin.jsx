import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import PropType from 'prop-types';

import FacebookButton from '../views/common/FacebookButton';
import GoogleButton from '../views/common/GoogleButton';
import { loginFacebook, loginGoogle }  from '../utils';
import socialLogin  from '../actions/socialAction';
import { globalError } from '../actions/globalActions';

/**
 * CLass representing Social Login (OAuth) Components
 *
 * @class SocialLogin
 * @extends {Component}
 */
export class SocialLogin extends Component {
  static propTypes = {
    handleLogin: PropType.func.isRequired,
    handleError: PropType.func.isRequired,
    isAuthenticated: PropType.bool.isRequired,
  }
  
  /**
   * Creates an instance of SocialLogin.
   * 
   * @returns {void}
   * @memberof SocialLogin
   */
  constructor() {
    super();
    this.socialResponse = this.socialResponse.bind(this);
  }

  /**
   * Handles response from service provider
   *
   * @param {object} response - The response from obtained facebook
   * @param {Function} callback - Response handler
   * @returns {void}
   */
  socialResponse(response, callback) {
    const { handleLogin, handleError } = this.props;
    callback(response, handleLogin, handleError);
  }

  /**
   * React render function
   *
   * @returns {JSX.Element} - DOM element
   * @memberof SocialLogin
   */
  render() {
    const { isAuthenticated } = this.props;
    if (isAuthenticated) { // redirect the user if authenticated
      return <Redirect to="/" />;
    }
    
    return (
      <div className="signup__socials__content">
        <div className="signup__socials__icons">
          <FacebookButton
            handleResponse={(response) => this.socialResponse(response, loginFacebook)}
          />
          <GoogleButton
            handleResponse={(response) => this.socialResponse(response, loginGoogle)}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  isAuthenticated: state.global.isAuthenticated,
});

const mapDispatchToProps = {
  handleLogin: (payload) => socialLogin(payload),
  handleError: (error) => globalError(error),
};

export default connect(mapStateToProps, mapDispatchToProps)(SocialLogin);
