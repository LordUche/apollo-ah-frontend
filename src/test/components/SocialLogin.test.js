/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { SocialLogin } from '../../components/SocialLogin';
import FacebookButton from '../../views/common/FacebookButton';
import GoogleButton from '../../views/common/GoogleButton';

/**
 * Wrapper for enzyme shallow component
 *
 * @returns {object} The mocked props and component
 */
function setup() {
  const props = {
    handleLogin: jest.fn(),
    handleError: jest.fn(),
    isAuthenticated: false,
  };

  const enzymeWrapper = shallow(<SocialLogin {...props} />);
  return { props, enzymeWrapper };
}

describe('<SocialLogin />', () => {  
  it('renders without crashing given the required props', () => {
    const { enzymeWrapper } = setup();
    expect(toJson(enzymeWrapper)).toMatchSnapshot();
  });

  it('should render FacebookButton', () => {
    const { enzymeWrapper } = setup();
    expect(enzymeWrapper.find(FacebookButton)).toHaveLength(1);
  });
  it('should render GoogleButton', () => {
    const { enzymeWrapper } = setup();
    expect(enzymeWrapper.find(GoogleButton)).toHaveLength(1);
  });
});
