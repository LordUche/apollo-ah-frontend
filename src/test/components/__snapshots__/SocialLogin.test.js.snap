// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SocialLogin /> renders without crashing given the required props 1`] = `
<div
  className="signup__socials__content"
>
  <div
    className="signup__socials__icons"
  >
    <FacebookButton
      handleResponse={[Function]}
    />
    <GoogleButton
      handleResponse={[Function]}
    />
  </div>
</div>
`;
